### PR TITLE
<PostCardList>

### PR DESCRIPTION
**Pull request by ryotogashi**

### CODE REVIEW

**Comment by ryotogashi** [comment]
@axross Can you check?

File: utilities/getPositionByPlayerAndIndex.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  SB = "SB",
+  BB = "BB",
+  UTG = "UTG",
+  EP2 = "EP2",
+  MP1 = "MP1",
+  MP2 = "MP2",
+  MP3 = "MP3",
+  LP1 = "LP1",
+  LP2 = "LP2",
+  BTN = "BTN"
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.ts]: Upper letters for enums are tricky. Use lower letters.

File: utilities/getPositionByPlayerAndIndex.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  playerLength: number,
+  index: number
+): Position {
+  if (playerLength < 2 || 10 < playerLength)
+    throw new Error("Player number must be 2 - 10.");
+  if (index < 0 || playerLength < index)
+    throw new Error(`Index number must be 0 - player number(${playerLength})`);
+
+  return playerNumberAndPositionCorrespondence.get(playerLength)![index];
+}
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.ts]: Nice! ✨

File: utilities/getPositionByPlayerAndIndex.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
+enum Position {
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.ts]: `Position` as a name sounds covering too much. Can imagine something like position on a display. How about `PlayerPosition`?

File: utilities/getSIMetricPrefixData.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    [7, "Z"],
+    [8, "Y"]
+  ]);
+  let data = num;
+  let i = 0;
+
+  while (1000 < data) {
+    data = data / 1000;
+    i++;
+  }
```
Comment (axross) [inline utilities/getSIMetricPrefixData.ts]: Using [`Math.log10()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/log10) is much faster and clear.

File: utilities/getAgoByDate.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+export default function getAgoByDate(date: Date): [number, string, string] {
```
Comment (axross) [inline utilities/getAgoByDate.ts]: Nice implementation. However, it would be better if this functions is split into two: long expression (for desktop) and short expression (for mobile).

Comment (axross) [reply utilities/getAgoByDate.ts]: Probably I prefer such name 🤔

```suggestion
export default function getRelativeDateString(date: Date): [number, string, string] {
```

File: components/PostCardListItem/PostCardListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+  max-width: 440px;
+  grid-template-rows: 1fr 0.25fr 0.25fr;
+  grid-template-columns: 1.3fr 0.7fr 0.9fr 0.9fr 1.2fr;
+  row-gap: 4px;
+  column-gap: 8px;
+`;
+
+const PlayerHandArea = styled.div`
+  grid-row: 1 / 2;
+  grid-column: 1 / 2;
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: Please define `grid-template-areas` at the parent and use `grid-area` at children. `grid-row` and `grid-column` have a huge problem at maintainability.

File: components/PostCardListItem/PostCardListItem.tsx [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        <TermSpan>{term}</TermSpan>
+        &nbsp;ago
+      </AttributeValue>
+    </Root>
+  );
+}
+
+const Root = styled(Card)`
+  display: grid;
+  max-width: 440px;
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: What's the intention of this `max-width`? Seems restricting usages.

File: components/PostCardListItem/PostCardListItem.tsx [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    : 0;
+  const turnPod = gameSituation.turn
+    ? gameSituation.turn.actions.reduce((sum, { betSize }) => sum + betSize, 0)
+    : 0;
+  const riverPod = gameSituation.river
+    ? gameSituation.river.actions.reduce((sum, { betSize }) => sum + betSize, 0)
+    : 0;
+
+  return preflopPod + flopPod + turnPod + riverPod;
+}
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: Nice!

File: components/PostCardListItem/PostCardListItem.tsx [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+`;
+
+const TermSpan = styled.span`
+  display: inline;
+  ${MOBILE_MEDIA} {
+    display: none;
+  }
+`;
+
+function getFinalPodOfTheGame(gameSituation: GameSituation): number {
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]:
```suggestion
function calculateFinalPot(gameSituation: GameSituation): number {
```

File: components/PostCardListItem/PostCardListItem.tsx [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    ? gameSituation.turn.actions.reduce((sum, { betSize }) => sum + betSize, 0)
+    : 0;
+  const riverPod = gameSituation.river
+    ? gameSituation.river.actions.reduce((sum, { betSize }) => sum + betSize, 0)
+    : 0;
+
+  return preflopPod + flopPod + turnPod + riverPod;
+}
+
+enum Round {
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: No one call "round" for those stuff. Use the term "street" instead.

(And please make sure whether you are using correct/official terms, EVERY TIME)

File: components/PostCardListItem/PostCardListItem.tsx [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+  }
+`;
+
+const AttributeTitle = styled.h5`
+  grid-row: 2 / 3;
+  margin: 0;
+  font-size: 14px;
+  color: #595959;
+  font-weight: normal;
+  ${MOBILE_MEDIA} {
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: Put a blank line. Because the second line will expand to be `@media screen and (...) {`.

```suggestion
  font-weight: normal;

  ${MOBILE_MEDIA} {
```

And so on below.

File: components/PostCardListItem/PostCardListItem.tsx [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+const HandCard = styled(PlayingCard)`
+  position: absolute;
+  width: 40%;
+
+  &:first-of-type {
+    top: 15%;
+    left 16%;
+    transform: rotate(-15deg);
+    -moz-transform: rotate(-15deg);
+    -webkit-transform: rotate(-15deg);
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: Seems you don't need to set vender prefixes. Styled Components automatically set them.

ref. https://styled-components.com/docs/basics#motivation

File: components/PostCardListItem/PostCardListItem.tsx [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+  post,
+  ...props
+}: Props) {
+  const { heroIndex, playerCards } = post.gameSituation;
+  const { left, right } = playerCards[heroIndex]!;
+  const [nAgo, mobileTerm, term] = getAgoByDate(
+    isRecentPost ? post.createdAt : post.lastUpdatedAt
+  );
+
+  return (
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]: JSX you wrote is a little bit untidy. Please refer others and clean it up.

File: components/PostCardListItem/PostCardListItem.tsx (hunk 9 above)
Comment (ryotogashi) [reply components/PostCardListItem/PostCardListItem.tsx]: I put this utilities folder since it will be used in new post page.
Screen Shot 2020-03-20 at 12 36 09 PM

**Comment by ryotogashi** [comment]
#93

File: components/PostCardListItem/PostCardListItem.tsx [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+        &nbsp;ago
+      </AttributeValue>
+    </Root>
+  );
+}
+
+const Root = styled(Card)`
+  display: grid;
+  max-width: 440px; // TODO
+  grid-template-rows: 1fr 0.25fr 0.25fr;
```
Comment (axross) [inline components/PostCardListItem/PostCardListItem.tsx]:
```suggestion
  grid-template-rows: 4fr 1fr 1fr;
```

**Comment by ryotogashi** [comment]
@axross Can you check when you have time?

File: components/PostCardList/PostCardList.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+export default {
+  title: "PostCardList",
+  component: PostCardList,
+  subcomponent: { PostCardListItem }
+};
+
+export const example = () => {
+  return (
+    <PostCardList postType={select("POST_TYPE", POST_TYPE, PostType.recent)}>
```
Comment (axross) [inline components/PostCardList/PostCardList.stories.tsx]: :+1:

```suggestion
    <PostCardList postType={select("postType", POST_TYPE, PostType.recent)}>
```

File: components/PostCardList/PostCardList.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+    <PostCardList postType={PostType.recent}>
+      <PostCardListItem post={post} onClick={action("onClick")} />
+    </PostCardList>
+    <PostCardList postType={PostType.lastUpdate}>
+      <PostCardListItem post={post} onClick={action("onClick")} />
+    </PostCardList>
+  </>
+);
+
+const POST_TYPE = {
```
Comment (axross) [inline components/PostCardList/PostCardList.stories.tsx]:
```suggestion
const POST_TYPES = {
```

File: components/PostCardList/PostCardList.stories.tsx [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    type: GameType.cash,
+    playerLength: 2,
+    playerStackSizes: [0, 0],
+    heroIndex: 0,
+    playerCards: [
+      {
+        left: {
+          rank: Rank.seven,
+          suit: Suit.heart
+        } as PlayingCard,
```
Comment (axross) [inline components/PostCardList/PostCardList.stories.tsx]: I'm not 100% sure but it looks you don't need `as PlayingCard` 🤔

```suggestion
        left: {
          rank: Rank.seven,
          suit: Suit.heart
        },
```

File: components/PostCardList/PostTypeContext.ts [hunk 18]
```diff
@@ -0,0 +1,8 @@
+import * as React from "react";
+
+export enum PostType {
+  recent,
+  lastUpdate
+}
```
Comment (axross) [inline components/PostCardList/PostTypeContext.ts]: What `PostType` means doesn't sound matching what it actually is. "PostType" sounds the type of post. Probably you need to come up with proper name for it.

File: components/PostCardList/PostCardList.tsx [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+      <PostTypeContext.Provider value={postType}>
+        {children}
+      </PostTypeContext.Provider>
+    </Root>
+  );
+}
+
+const Root = styled.div`
+  display: grid;
+  grid-template-columns: 440px 440px;
```
Comment (axross) [inline components/PostCardList/PostCardList.tsx]: Please make components responsible. What it will show if the screen width is 960px, 1280px, or 1440px?

File: components/PostCardList/PostCardListItem.tsx [hunk 20]
```diff
@@ -0,0 +1,219 @@
+import * as React from "react";
+import styled from "styled-components";
+import { MOBILE_MEDIA } from "../../constants/mediaquery";
+import Post from "../../models/Post";
+import calculateFinalPot from "../../utilities/calculateFinalPot";
+import getAgoByDate from "../../utilities/getRelativeDateString";
+import getPositionByPlayerAndIndex from "../../utilities/getPositionByPlayerAndIndex";
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]:
```suggestion
import getPositionByPlayerAndIndex from "../../utilities/getPositionByPlayerAndIndex";
import getRelativeDateString from "../../utilities/getRelativeDateString";
```

File: components/PostCardList/PostCardListItem.tsx [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+  style?: React.CSSProperties;
+}
+
+export default function PostCardListItem({ post, ...props }: Props) {
+  const { heroIndex, playerCards } = post.gameSituation;
+  const { left, right } = playerCards[heroIndex]!;
+  const postType = React.useContext(PostTypeContext);
+  const [nAgo, mobileTerm, term] = getAgoByDate(
+    postType === PostType.recent ? post.createdAt : post.lastUpdatedAt
+  );
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]: This function's return value signature is quite tricky. Can you tell what `nAgo`, `mobileTerm` and `term` is, in an instant in the future?

File: components/PostCardList/PostCardListItem.tsx [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+        <Attribute>
+          <h5>Play at</h5>
+          <span>
+            {post.gameSituation.river
+              ? Street.river
+              : post.gameSituation.turn
+              ? Street.turn
+              : post.gameSituation.flop
+              ? Street.flop
+              : Street.preflop}
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]: This is tricky too.

File: components/PostCardList/PostCardListItem.tsx [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+        </Attribute>
+      </Attributes>
+    </Root>
+  );
+}
+
+const Root = styled(Card)`
+  display: grid;
+  grid-template-rows: 2fr 1fr;
+  grid-template-columns: 1.3fr 3.7fr;
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]:
```suggestion
  grid-template-columns: 13fr 37fr;
```

File: utilities/calculateFinalPot.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
+import GameSituation from "../models/GameSituation";
+
+export default function calculateFinalPot(
+  gameSituation: GameSituation
+): number {
+  const preflopPod = gameSituation.preflop.actions.reduce(
```
Comment (axross) [inline utilities/calculateFinalPot.ts]: Not "pod", it's "pot".

File: utilities/getPositionByPlayerAndIndex.ts [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+  lp2 = "LP2",
+  btn = "BTN"
+}
+
+export default function getPositionByPlayerAndIndex(
+  playerLength: number,
+  index: number
+): PlayerPosition {
+  if (playerLength < 2 || 10 < playerLength)
+    throw new Error("Player number must be 2 - 10.");
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.ts]:
```suggestion
    throw new Error("playerLength must be greater than or equal 2 and less than or equal 10.");
```

File: utilities/getRelativeDateString.ts [hunk 26]
```diff
@@ -0,0 +1,26 @@
+export default function getRelativeDateString(
+  date: Date
+): [number, string, string] {
+  const now = new Date();
+  const yearDiff = now.getFullYear() - date.getFullYear();
```
Comment (axross) [inline utilities/getRelativeDateString.ts]:
```suggestion
  const now = new Date();

  const yearDiff = now.getFullYear() - date.getFullYear();
```

**Comment by ryotogashi** [comment]
@axross Thank you for your review I appreciate it! I fix them. Can you check again when you have time?

File: components/PostCardList/PostCardListItem.tsx [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+  style?: React.CSSProperties;
+}
+
+export default function PostCardListItem({ post, ...props }: Props) {
+  const { heroIndex, playerCards } = post.gameSituation;
+  const { left, right } = playerCards[heroIndex]!;
+  const postType = React.useContext(PostTypeContext);
+  const [date, mobileDateUnit, dateUnit] = getRelativeDateString(
+    postType === PostType.recent ? post.createdAt : post.lastUpdatedAt
+  );
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]: It would be better if you split it into two functions, such like `getRelativeDateString()` and `getRelativeShortDateString`

File: components/PostCardList/PostCardListItem.tsx [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+    </Root>
+  );
+}
+
+const Root = styled(Card)`
+  display: grid;
+  grid-template-columns: minmax(15%, 100px) 1fr;
+  grid-template-areas:
+    "playinghand-area title-area"
+    "likes-area attributes-area";
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]:
```suggestion
    "playing-cards title"
    "likes attributes";
```

Comment (axross) [reply components/PostCardList/PostCardListItem.tsx]: Use proper terms. Please.

File: components/PostCardList/PostCardListItem.tsx [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+      </Attributes>
+    </Root>
+  );
+}
+
+const Root = styled(Card)`
+  display: grid;
+  grid-template-columns: minmax(15%, 100px) 1fr;
+  grid-template-areas:
+    "playinghand-area title-area"
```
Comment (axross) [inline components/PostCardList/PostCardListItem.tsx]: Use proper terms. Please.

If you are not sure, refer to dictionaries or [google](https://www.google.com/search?q=poker+playing+hand).

File: utilities/calculateFinalPot.test.ts [hunk 30]
```diff
@@ -0,0 +1,190 @@
...
+        {
+          playerIndex: 1,
+          betSize: 0
+        }
+      ]
+    }
+  };
+
+  it("returns final pot when ended street is preflop", () => {
+    expect(calculateFinalPot(baseGameSituationData)).toBe(0.025);
```
Comment (axross) [inline utilities/calculateFinalPot.test.ts]: The same thing happened with https://github.com/agnosticful/chipstackoverflow-web/pull/128#discussion_r396835168

File: utilities/getPositionByPlayerAndIndex.test.ts [hunk 31]
```diff
@@ -0,0 +1,127 @@
+import getPositionByPlayerAndIndex from "./getPositionByPlayerAndIndex";
+
+describe("getPositionByPlayerAndIndex()", () => {
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.test.ts]: Nice! ✨

File: utilities/getPositionByPlayerAndIndex.test.ts [hunk 32]
```diff
@@ -0,0 +1,127 @@
...
+  it("should throw playerLength error", () => {
+    function playerLengthTest1() {
+      getPositionByPlayerAndIndex(1, 0);
+    }
+
+    expect(playerLengthTest1).toThrowError(
+      new Error(
+        "playerLength must be greater than or equal 2 and less than or equal 10."
+      )
+    );
```
Comment (axross) [inline utilities/getPositionByPlayerAndIndex.test.ts]:
```suggestion
    expect(() => getPositionByPlayerAndIndex(1, 0)).toThrowError(
      new Error(
        "playerLength must be greater than or equal 2 and less than or equal 10."
      )
    );
```

File: utilities/getRelativeDateString.test.ts [hunk 33]
```diff
@@ -0,0 +1,87 @@
+import getRelativeDateString from "./getRelativeDateString";
+
+describe("getRelativeDateString()", () => {
+  it("returns 1 year", () => {
+    const now = new Date();
+    now.setFullYear(now.getFullYear() - 1);
+    expect(getRelativeDateString(now)).toEqual([1, "y", "year"]);
```
Comment (axross) [inline utilities/getRelativeDateString.test.ts]:
```suggestion
    const now = new Date();
    now.setFullYear(now.getFullYear() - 1);

    expect(getRelativeDateString(now)).toEqual([1, "y", "year"]);
```

File: utilities/getSIMetricPrefixData.test.ts [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+    expect(getSIMetricPrefixData(500000000000000000000)).toBe("500E");
+    expect(getSIMetricPrefixData(5000000000000000000000)).toBe("5Z");
+    expect(getSIMetricPrefixData(50000000000000000000000)).toBe("50Z");
+    expect(getSIMetricPrefixData(500000000000000000000000)).toBe("500Z");
+    expect(getSIMetricPrefixData(5000000000000000000000000)).toBe("5Y");
+    expect(getSIMetricPrefixData(50000000000000000000000000)).toBe("50Y");
+    expect(getSIMetricPrefixData(500000000000000000000000000)).toBe("500Y");
+  });
+
+  it("works edge case well", () => {
```
Comment (axross) [inline utilities/getSIMetricPrefixData.test.ts]:
```suggestion
  it("works for an edge case well", () => {
```

File: utilities/getSIMetricPrefixData.test.ts [hunk 35]
```diff
@@ -0,0 +1,40 @@
+import getSIMetricPrefixData from "./getSIMetricPrefixData";
+
+describe("getSIMetricPrefixData()", () => {
+  it("returns number without prefix", () => {
```
Comment (axross) [inline utilities/getSIMetricPrefixData.test.ts]:
```suggestion
  it("returns string of number", () => {
```

File: utilities/getSIMetricPrefixData.test.ts [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+    expect(getSIMetricPrefixData(5)).toBe("5");
+    expect(getSIMetricPrefixData(50)).toBe("50");
+    expect(getSIMetricPrefixData(500)).toBe("500");
+  });
+
+  it("returns number with prefix", () => {
```
Comment (axross) [inline utilities/getSIMetricPrefixData.test.ts]:
```suggestion
  it("returns string of number with suffix", () => {
```

File: utilities/getSIMetricPrefixData.ts [hunk 37]
```diff
@@ -0,0 +1,27 @@
+export default function getSIMetricPrefixData(num: number): string {
```
Comment (axross) [inline utilities/getSIMetricPrefixData.ts]: Keep in mind: do not use "data" for identifiers. "data" is unclear. it means nothing. So `getSIMetricPrefixData` means the same with `getSIMetricPrefix`. If you wanna intend something, define it clearly. Probably `String`, `DataString` or things like that. You will **never** face with appropriate situation to use `xxxData` in your developer life.

Comment (axross) [reply utilities/getSIMetricPrefixData.ts]: By the way, it should be "suffix" yet "prefix", I guess.

File: utilities/calculateFinalPot.test.ts (hunk 30 above)
Comment (ryotogashi) [reply utilities/calculateFinalPot.test.ts]: @axross Thank you for your comment. However, baseGameSituationData is used twice actually. ( It used at line 41 and 176). Only gameSituationData3 was used once and I knew it but I did it for readability.  What do you think?

File: utilities/getPositionByPlayerAndIndex.test.ts (hunk 32 above)
Comment (ryotogashi) [reply utilities/getPositionByPlayerAndIndex.test.ts]: omg. That is true... much better.

File: utilities/calculateFinalPot.test.ts (hunk 30 above)
Comment (axross) [reply utilities/calculateFinalPot.test.ts]: I’m sorry I overlooked it. Then please only change the identifier to be proper one.

Comment (ryotogashi) [reply utilities/calculateFinalPot.test.ts]: Thanks

**Comment by ryotogashi** [comment]
I'll request another one. Thanks
